src/main/java/de/medizininformatik_initiative/processes/common/fhir/client/token/OAuth2TokenClient.java

```diff
@@ -0,0 +1,231 @@
+		}
+		catch (IOException | InterruptedException exception)
+		{
+			throw new RuntimeException(exception);
+		}
+	}
+
+	private HttpClient createClient()
+	{
+		HttpClient.Builder builder = HttpClient.newBuilder();
+		builder.connectTimeout(Duration.ofMillis(connectTimeout));
+
+		configureProxy(builder);
+		configureTrustStore(builder);
+
+		return builder.build();
+	}
+
+	private void configureTrustStore(HttpClient.Builder builder)
+	{
+		if (trustStorePath != null)
+		{
+			logger.debug("Reading trust-store from {}", trustStorePath.toString());
+			KeyStore trustStore = readTrustStore(trustStorePath);
+			SSLContext sslContext = createSslContext(trustStore);
+			builder.sslContext(sslContext);
+		}
+	}
+
+	private SSLContext createSslContext(KeyStore trustStore)
+	{
+		try
+		{
+			TrustManagerFactory trustManagerFactory = TrustManagerFactory
+					.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+			trustManagerFactory.init(trustStore);
+
+			SSLContext sslContext = SSLContext.getInstance(TLS);
+			sslContext.init(null, trustManagerFactory.getTrustManagers(), null);
+
+			return sslContext;
+		}
+		catch (NoSuchAlgorithmException | KeyStoreException | KeyManagementException exception)
+		{
+			throw new RuntimeException(exception);
+		}
+	}
+
+	private KeyStore readTrustStore(Path trustPath)
+	{
+		try
+		{
+			return CertificateReader.allFromCer(trustPath);
+		}
+		catch (NoSuchAlgorithmException | CertificateException | KeyStoreException | IOException exception)
+		{
+			throw new RuntimeException(exception);
+		}
+	}
+
+	private void configureProxy(HttpClient.Builder builder)
+	{
+		if (proxyUrl != null)
+		{
+			URI uri = URI.create(proxyUrl);
+			builder.proxy(ProxySelector.of(new InetSocketAddress(uri.getHost(), uri.getPort())));
+		}
+	}
+
+	private HttpRequest createAccessTokenRequest()
+	{
+		HttpRequest.Builder builder = HttpRequest.newBuilder();
+		builder.uri(URI.create(issuerUrl));
+		builder.timeout(Duration.ofMillis(socketTimeout));
+
+		configureAuthentication(builder);
+		configureProxyAuthentication(builder);
+
+		builder.header(HEADER_CONTENT_TYPE, MIME_TYPE_FORM_URLENCODED);
+		builder.POST(HttpRequest.BodyPublishers.ofString(GRANT_TYPE_CLIENT_CREDENTIALS));
+
+		return builder.build();
+	}
+
+	private void configureAuthentication(HttpRequest.Builder builder)
+	{
+		// Keycloak not sending WWW-Authenticate header for response code 401
+		String credentials = getCredentials(clientId, clientSecret);
+		builder.header(HEADER_AUTHORIZATION, PREFIX_BASIC_AUTH + credentials);
+	}
+
+	private void configureProxyAuthentication(HttpRequest.Builder builder)
+	{
+		// Proxy authentication using similar workaround as basic authentication
+		if (proxyUrl != null && proxyUsername != null && proxyPassword != null)
+		{
+			String proxyCredentials = getCredentials(proxyUsername, proxyPassword);
+			builder.setHeader(HEADER_AUTHORIZATION_PROXY, PREFIX_BASIC_AUTH + proxyCredentials);
+		}
+	}
```

Comment:
Do not use preemptive basic auth for proxy. Proxy always sends challenge and may not use basic auth. See RFC 2617 HTTP Authentication: Basic and Digest Access Authentication section [3.6 Proxy-Authentication and Proxy-Authorization](https://datatracker.ietf.org/doc/html/rfc2617#section-3.6).
```suggestion
		if (proxyUrl != null && proxyUsername != null && proxyPassword != null)
		{
			Authenticator a = new Authenticator()
			{
				@Override
				protected PasswordAuthentication getPasswordAuthentication()
				{
					return new PasswordAuthentication(proxyUsername, proxyPassword);
				}
			};
			builder.authenticator(a);
		}
	}
```

Reply:
Not setting preemptive proxy authorization overrides preemptive authorization for oauth2 provider.

Reply:
Probably related to https://bugs.openjdk.org/browse/JDK-8326949